lib/AST/ArchetypeBuilder.cpp

```diff
@@ -1002,8 +1007,14 @@ bool ArchetypeBuilder::addConformanceRequirement(PotentialArchetype *PAT,
   if (!T->addConformance(Proto, /*updateExistingSource=*/true, Source, *this))
     return false;
 
-  RequirementSource InnerSource(RequirementSource::Redundant, Source.getLoc());
-  
+  // Conformances to inherit protocols are explicit in a protocol requirement
+  // signature, but inferred from this conformance otherwise.
+  auto InnerKind =
+      Source.getKind() == RequirementSource::ProtocolRequirementSignatureSelf
+          ? RequirementSource::Explicit
+          : RequirementSource::Redundant;
```

Comment:
I know this is just a refactoring of what was there before, but it seems odd to use `RequirementSource::Redundant` here rather than `RequirementSource::Protocol`. Maybe we should (separately!) see if we can make that change to align it with what we do below.